service/history/workflow/task_generator.go

```diff
@@ -694,14 +694,16 @@ func (r *TaskGeneratorImpl) GenerateUserTimerTasks() error {
 }
 
 func (r *TaskGeneratorImpl) GenerateHistoryReplicationTasks(
-	events []*historypb.HistoryEvent,
+	eventBatches [][]*historypb.HistoryEvent,
 ) error {
-	if len(events) == 0 {
+	if len(eventBatches) == 0 {
 		return nil
 	}
 
-	firstEvent := events[0]
-	lastEvent := events[len(events)-1]
+	firstBatch := eventBatches[0]
+	firstEvent := firstBatch[0]
+	lastBatch := eventBatches[len(eventBatches)-1]
+	lastEvent := lastBatch[len(lastBatch)-1]
```

Comment:
I think we need the first event version abd last event version to locate which history branch to use

Reply:
Or at least the version should be last event version to locate the history branch.

Reply:
Synced offline. They will be generated from one transaction. Could we add a sanity check to make sure they use the same version?

Reply:
check added.